File: 2594. Minimum Time to Repair Cars/2594.myans.cpp
class Solution {
    bool canRepairCarsInTime(vector<int>& ranks, int cars, long long time) {
        int sum = 0;
        for (int& rank : ranks) {
            sum += std::sqrt(time / rank);
            if (sum >= cars) return true;
        }
        return false;
    }

public:
    long long repairCars(vector<int>& ranks, int cars) {
        // 最小のrankを求める
        int minRank = *std::min_element(ranks.begin(), ranks.end());
        // ranksが空配列になり、rank.end()が返ってくることは、今回は制約上ないため、例外処理はしない

        // 最小のrankのみが行った場合の時間を最大値とし、二分探索で最小値を求める
        long long left = 0, right = static_cast<long long>(minRank)*cars*cars;
        while (left < right) {
            long long middle = (left + right) / 2;
            if (canRepairCarsInTime(ranks, cars, middle)) {
                right = middle;
            } else {
                left = middle + 1;
            }
        }

        return left;
    }
};

File: 2594. Minimum Time to Repair Cars/2594.myans.rs
impl Solution {
    fn canRepairCarsInTime(ranks: &[i32], cars: i32, time: i64) -> bool {
        let mut sum = 0;
        for &rank in ranks.iter() {
            sum += ((time / rank as i64) as f64).sqrt() as i32;
            if sum >= cars {
                return true
            }
        }
        false
    }

    pub fn repair_cars(ranks: Vec<i32>, cars: i32) -> i64 {
        // rankの最小値を取る
        let min_rank = *ranks.iter().min().unwrap();

        let mut left = 0;
        let mut right = min_rank as i64 * cars as i64 * cars as i64;
        while left < right {
            let middle = left + (right - left) / 2;
            if Self::canRepairCarsInTime(&ranks, cars, middle) {
                right = middle;
            } else {
                left = middle + 1;
            }
        }
        left
    }
}